---
title: 'Introduction to BASiCS: Bayesian Analysis of Single-Cell Sequencing data (no spike case)'
author: "Catalina A. Vallejos"
date: '`r Sys.Date()`'
output: html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title} 
  %\VignetteEngine{knitr::rmarkdown} 
  \usepackage[utf8]{inputenc}
---

***

# Introduction 

Here, we demonstrate how to use BASiCS using a simulated dataset that does not contain spike in genes. 

>> THIS FUNCTIONALITY IS UNDER DEVELOPMENT. NOT READY TO BE USED.

***

# Installation

BASiCS can be installed directly from Github. 

```{r install, include=FALSE}
#library(devtools)
#install_github('catavallejos/BASiCS', build_vignettes = TRUE)
```

This installation might fail if some of the dependency libraries are not yet installed. If so, please run the following lines and re-install. 

```{r dependencies}
#library(devtools)
#source("http://bioconductor.org/biocLite.R")
#biocLite("BiocGenerics")
#install.packages("Rcpp")
```

After a successful installation, the BASiCS library can be loaded using[^footnoteInstall] 

```{r load_packages}
library(BASiCS)
```

[^footnoteInstall]: The warning `"No methods found in "BiocGenerics""` might appear. Please ignore. `BASiCS` imports some of the generic functions provided by `BiocGenerics` that do not have any methods attached.

# Example

##  The input dataset

For illustration purposes, the BASiCS package includes a built-in simulated dataset:

```{r ExampleData}
Data = makeExampleBASiCS_Data(WithSpikes = FALSE)
```

Which was generated using the following simulated parameters

```{r SimulatedParam}
Mu =  c( 8.36,  10.65,   4.88,   6.29,  21.72,  12.93,  30.19,  83.92,   3.89,   6.34,
           57.87,  12.45,   8.08,   7.31,  15.56,  15.91,  12.24,  15.96,  19.14,   4.20,
           6.75,  27.74,   8.88,  21.21,  19.89,   7.14,  11.09,   7.19,  20.64,  73.90,
           9.05,   6.13,  16.40,   6.08,  17.89,   6.98,  10.25,  14.05,   8.14,   5.67,
           6.95,  11.16,  11.83,   7.56, 159.05,  16.41,   4.58,  15.46,  10.96,  25.05,
           18.54,   8.50,   4.05,   5.37,   4.63,   4.08,   3.75,   5.02,  27.74,  10.28,
           3.91,  13.10,   8.23,   3.64,  80.77,  20.36,   3.47,  20.12,  13.29,   7.92,
           25.51, 173.01,  27.71,   4.89,  33.10,   3.42,   6.19,   4.29,   5.19,   8.36,
           10.27,   6.86,   5.72,  37.25,   3.82,  23.97,   5.80,  14.30,  29.07,   5.30,
           7.47,   8.44,   4.24,  16.15,  23.39, 120.22,   8.92,  97.15,   9.75,  10.07)
Delta = c(1.29, 0.88, 1.51, 1.49, 0.54, 0.40, 0.85, 0.27, 0.53, 1.31,
            0.26, 0.81, 0.72, 0.70, 0.96, 0.58, 1.15, 0.82, 0.25, 5.32,
            1.13, 0.31, 0.66, 0.27, 0.76, 1.39, 1.18, 1.57, 0.55, 0.17,
            1.40, 1.47, 0.57, 2.55, 0.62, 0.77, 1.47, 0.91, 1.53, 2.89,
            1.43, 0.77, 1.37, 0.57, 0.15, 0.33, 3.99, 0.47, 0.87, 0.86,
            0.97, 1.25, 2.20, 2.19, 1.26, 1.89, 1.70, 1.89, 0.69, 1.63,
            2.83, 0.29, 1.21, 2.06, 0.20, 0.34, 0.71, 0.61, 0.71, 1.20,
            0.88, 0.17, 0.25, 1.48, 0.31, 2.49, 2.75, 1.43, 2.65, 1.97,
            0.84, 0.81, 2.75, 0.53, 2.23, 0.45, 1.87, 0.74, 0.53, 0.58,
            0.94, 0.72, 2.61, 1.56, 0.37, 0.07, 0.90, 0.12, 0.76, 1.45)
Phi = c(1.09, 1.16, 1.19, 1.14, 0.87, 1.10, 0.48, 1.06, 0.94, 0.97,
          1.09, 1.16, 1.19, 1.14, 0.87, 1.10, 0.48, 1.06, 0.94, 0.97)
Phi[1:10] = 2 * Phi[1:10]
Theta1 = 0.5; Theta2 = 0.75;
Theta = ifelse(1:20 <= 10, Theta1, Theta2)
set.seed(1000); Nu<-rgamma(20,shape=1/Theta,rate=1/(Phi*Theta))
```

## Fitting the BASiCS model

```{r MCMC}
# The required number of iterations to reach convergence varies for different datasets
Start = BASiCS:::HiddenBASiCS_MCMC_Start(Data)
#Start$delta0 <- Delta
#Start$phi0 <- Phi
#Start$theta0 <- (Theta1 + Theta2)/2
#Start$nu0 <- Nu
#Start$ls.mu0 = rep(11, times = 100)
#Start$ls.mu0 = log(log(Start$mu0)) - 2
#PriorParam = list(s2.mu = 0.25, s2.delta = 0.5, a.delta = 1, b.delta = 1, p.phi = rep(1, times = 20), a.phi = 1, b.phi = 1, a.s = 1, b.s = 1, a.theta = 1, b.theta = 1)
MCMC_Output <- BASiCS_MCMC(Data, N = 20000, Thin = 2, Burn = 10000, 
                           PrintProgress = TRUE, ConstrainType = 2) #PriorParam = PriorParam Start = Start, 
MCMC_Summary <- Summary(MCMC_Output)
```


```{r Traceplots, fig.width = 8, fig.height=5, message=FALSE, warning=FALSE}
plot(MCMC_Output, Param = "mu", Gene = 1, log = "y")
abline(h = Mu[1], col = "red")
plot(MCMC_Output, Param = "mu", Gene = 5, log = "y")
abline(h = Mu[5], col = "red")
plot(MCMC_Output, Param = "mu", Gene = 10, log = "y")
abline(h = Mu[10], col = "red")
plot(MCMC_Output, Param = "mu", Gene = 51, log = "y")
abline(h = Mu[51], col = "red")
plot(MCMC_Output, Param = "mu", Gene = 100, log = "y")
abline(h = Mu[100], col = "red")
plot(MCMC_Output, Param = "delta", Gene = 1)
abline(h = Delta[1], col = "red")
plot(MCMC_Output, Param = "nu", Cell = 1)
abline(h = Nu[1], col = "red")
plot(MCMC_Output, Param = "phi", Cell = 1)
abline(h = Phi[1], col = "red")
plot(MCMC_Output, Param = "theta", Batch = 1) 
abline(h = Theta1, col = "red")
plot(MCMC_Output, Param = "theta", Batch = 2) 
abline(h = Theta2, col = "red")
```

```{r HPDplots, fig.width=8, fig.height=5, message=FALSE, warning=FALSE}
plot(MCMC_Summary, Param = "mu", log = "y")
points(Mu, pch = 16, col = "red")
plot(Mu, MCMC_Summary@mu[,1], log = "xy", pch = 16, col = 8, bty = "n")
abline(a = 0, b = 1)
plot(MCMC_Summary, Param = "delta")
points(Delta, pch = 16, col = "red")
plot(Delta, MCMC_Summary@delta[,1], log = "xy", pch = 16, col = 8, bty = "n")
abline(a = 0, b = 1)
plot(MCMC_Summary, Param = "phi")
points(Phi, pch = 16, col = "red")
plot(Phi, MCMC_Summary@phi[,1], pch = 16, col = 8, bty = "n")
abline(a = 0, b = 1)
plot(MCMC_Summary, Param = "nu")
points(Nu, pch = 16, col = "red")
plot(Nu, MCMC_Summary@nu[,1], pch = 16, col = 8, bty = "n")
abline(a = 0, b = 1)
plot(MCMC_Summary, Param = "theta")
points(c(Theta1, Theta2), pch = 16, col = "red")
```

