#' @name makeExampleBASiCS_Data
#'
#' @title Create a simple example of a SummarizedExperiment object with random data
#'
#' @description A simple \code{\linkS4class{SummarizedExperiment}} object is generated by simulating a dataset from the
#' model in BASiCS (Vallejos et al 2015). This is used for the examples in the package and the vignette.
#'
#' @param WithBatch If true, 2 batches are generated (each of them containing 10 cells)
#' @param WithSpikes If true, the simulated dataset contains 20 spike-in genes
#'
#' @return An object of class \code{\linkS4class{SummarizedExperiment}}, simulated from the model implemented in BASiCS.
#' It contains 120 genes (100 biological and 20 spike-in) and 20 cells.
#'
#' @examples
#' Data = makeExampleBASiCS_Data()
#' is(Data, "SummarizedExperiment")
#'
#' @author Catalina A. Vallejos \email{cnvallej@@uc.cl}
#'
#' @references Vallejos, Marioni and Richardson (2015). Bayesian Analysis of Single-Cell Sequencing data.
makeExampleBASiCS_Data <- function(WithBatch = FALSE, WithSpikes = TRUE)
{
  Mu =  c( 8.36,  10.65,   4.88,   6.29,  21.72,  12.93,  30.19,  83.92,   3.89,   6.34,
           57.87,  12.45,   8.08,   7.31,  15.56,  15.91,  12.24,  15.96,  19.14,   4.20,
           6.75,  27.74,   8.88,  21.21,  19.89,   7.14,  11.09,   7.19,  20.64,  73.90,
           9.05,   6.13,  16.40,   6.08,  17.89,   6.98,  10.25,  14.05,   8.14,   5.67,
           6.95,  11.16,  11.83,   7.56, 159.05,  16.41,   4.58,  15.46,  10.96,  25.05,
           18.54,   8.50,   4.05,   5.37,   4.63,   4.08,   3.75,   5.02,  27.74,  10.28,
           3.91,  13.10,   8.23,   3.64,  80.77,  20.36,   3.47,  20.12,  13.29,   7.92,
           25.51, 173.01,  27.71,   4.89,  33.10,   3.42,   6.19,   4.29,   5.19,   8.36,
           10.27,   6.86,   5.72,  37.25,   3.82,  23.97,   5.80,  14.30,  29.07,   5.30,
           7.47,   8.44,   4.24,  16.15,  23.39, 120.22,   8.92,  97.15,   9.75,  10.07,
           1010.72,   7.90,  31.59,  63.17,   1.97, 252.68,  31.59,  31.59,  31.59,  63.17,
           4042.89,4042.89,   3.95, 126.34, 252.68,   7.90,  15.79, 126.34,   3.95,  15.79)
  Delta = c(1.29, 0.88, 1.51, 1.49, 0.54, 0.40, 0.85, 0.27, 0.53, 1.31,
            0.26, 0.81, 0.72, 0.70, 0.96, 0.58, 1.15, 0.82, 0.25, 5.32,
            1.13, 0.31, 0.66, 0.27, 0.76, 1.39, 1.18, 1.57, 0.55, 0.17,
            1.40, 1.47, 0.57, 2.55, 0.62, 0.77, 1.47, 0.91, 1.53, 2.89,
            1.43, 0.77, 1.37, 0.57, 0.15, 0.33, 3.99, 0.47, 0.87, 0.86,
            0.97, 1.25, 2.20, 2.19, 1.26, 1.89, 1.70, 1.89, 0.69, 1.63,
            2.83, 0.29, 1.21, 2.06, 0.20, 0.34, 0.71, 0.61, 0.71, 1.20,
            0.88, 0.17, 0.25, 1.48, 0.31, 2.49, 2.75, 1.43, 2.65, 1.97,
            0.84, 0.81, 2.75, 0.53, 2.23, 0.45, 1.87, 0.74, 0.53, 0.58,
            0.94, 0.72, 2.61, 1.56, 0.37, 0.07, 0.90, 0.12, 0.76, 1.45)
  Phi = c(1.09, 1.16, 1.19, 1.14, 0.87, 1.10, 0.48, 1.06, 0.94, 0.97,
          1.09, 1.16, 1.19, 1.14, 0.87, 1.10, 0.48, 1.06, 0.94, 0.97)
  S = c(0.38, 0.40, 0.38, 0.39, 0.34, 0.39, 0.31, 0.39, 0.40, 0.37,
        0.38, 0.40, 0.38, 0.39, 0.34, 0.39, 0.31, 0.39, 0.40, 0.37)
  
  q = length(Mu); q.bio = length(Delta); n = length(Phi)
  
  if(!WithBatch) {Theta = 0.5}
  else
  {
    # 2 batches, 10 cells each
    Theta1 = 0.5; Theta2 = 0.75;
    Theta = ifelse(1:n <= 10, Theta1, Theta2)
  }
  
  if(WithSpikes)
  {
    # Matrix where simulated counts will be stored
    Counts.sim<-matrix(0,ncol=n,nrow=q)
    # Matrix where gene-cell specific simulated random effects will be stored
    Rho<-matrix(1,ncol=n,nrow=q.bio)
    # Simulated cell-specific random effects
    if(all(Theta>0)) {set.seed(1000); Nu<-rgamma(n,shape=1/Theta,rate=1/(S*Theta))}
    else {Nu<-S}
    # Simulated counts data
    for(i in 1:q)
    {
      # Biological genes
      if(i<=q.bio)
      {
        if(Delta[i]>0){set.seed(i); Rho[i,]<-rgamma(n,shape=1/Delta[i],rate=1/Delta[i])}
        set.seed(i+10000); Counts.sim[i,]<-rpois(n,lambda=Phi*Nu*Rho[i,]*Mu[i])
      }
      # Technical genes
      else {set.seed(i+20000); Counts.sim[i,]<-rpois(n,lambda=Nu*Mu[i])}
    }
    rownames(Counts.sim) <- paste0("Gene", 1:q)
    SpikeInfo = data.frame(paste0("Gene", (q.bio+1):q), Mu[(q.bio+1):q])
    
    if(!WithBatch)
    {
      Data = new("BASiCS_Data", Counts = Counts.sim, Tech = ifelse(1:q > q.bio, T, F), SpikeInput = SpikeInfo[,2], GeneNames = paste0("Gene", 1:q), BatchInfo = rep(1,20))
    }
    else
    {
      Data = new("BASiCS_Data", Counts = Counts.sim, Tech = ifelse(1:q > q.bio, T, F), SpikeInput = SpikeInfo[,2], GeneNames = paste0("Gene", 1:q),
                 BatchInfo = c(rep(1,10), rep(2,10)))
    }
    
  }
  else
  {
    # Matrix where simulated counts will be stored
    Counts.sim<-matrix(0,ncol=n,nrow=q.bio)
    # Matrix where gene-cell specific simulated random effects will be stored
    Rho<-matrix(1,ncol=n,nrow=q.bio)
    # Simulated cell-specific random effects
    Phi[1:10] = 2 * Phi[1:10]
    if(all(Theta>0)) {set.seed(1000); Nu<-rgamma(n,shape=1/Theta,rate=1/(Phi*Theta))}
    else {Nu<-Phi}
    # Simulated counts data
    for(i in 1:q.bio)
    {
      if(Delta[i]>0){set.seed(i); Rho[i,]<-rgamma(n,shape=1/Delta[i],rate=1/Delta[i])}
      set.seed(i+10000); Counts.sim[i,]<-rpois(n,lambda=Nu*Rho[i,]*Mu[i])
    }
    rownames(Counts.sim) <- paste0("Gene", 1:q.bio)
    
    Data = newBASiCS_Data(Counts = Counts.sim, Tech = rep(FALSE, q.bio), SpikeInput = 1, 
               BatchInfo = c(rep(1,10), rep(2,10)))
    
  }
  
  #  nBatch = length(unique(Data@BatchInfo))
  #  cat("An object of class ", class(Data), "\n", sep = "")
  #  cat(" Dataset contains ", q, " genes (", q.bio, " biological and ", q-q.bio, " technical) and ", n, " cells.\n", sep="")
  #  cat(" Elements (slots): Counts, Tech, SpikeInput, GeneNames and BatchInfo.\n")
  #  if(nBatch == 1) {cat(paste0(" The data contains ",nBatch," batch.\n"))}
  #  else {cat(paste0(" The data contains ",nBatch," batches.\n"))} 
  
  return(Data)
}